src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,147 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final String INDEX_OUT_OF_BOUNDS_MASSAGE
+            = "The index value is outside the array";
+    private static final int DEFAULT_SIZE_OF_ARRAY = 10;
+    private int sizeOfArray;
+    private T[] array;
+    private T[] temporaryArray;
+
+    public ArrayList() {
+        array = (T[])(new Object[DEFAULT_SIZE_OF_ARRAY]);
+    }
+
     @Override
     public void add(T value) {
-
+        if (sizeOfArray == array.length) {
+            temporaryArray = grow(sizeOfArray);
+            for (int i = 0; i < array.length; i++) {
+                temporaryArray[i] = array[i];
+            }
+            array = temporaryArray;
+        }
+        array[sizeOfArray] = value;
+        sizeOfArray++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > sizeOfArray || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MASSAGE);
+        }
+        if (sizeOfArray == array.length) {
+            temporaryArray = grow(sizeOfArray);
+        } else {
+            temporaryArray = (T[])(new Object[array.length]);
+        }
+        for (int i = 0; i < index; i++) {
+            temporaryArray[i] = array[i];
+        }
+        temporaryArray[index] = value;
+        sizeOfArray++;
+        for (int i = ++index; i < sizeOfArray; i++) {
+            temporaryArray[i] = array[i - 1];
+        }
+        array = temporaryArray;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int indexOfArrayList = 0;
+        if ((sizeOfArray + list.size()) >= array.length) {
+            temporaryArray = grow(sizeOfArray + list.size());
+        } else {
+            temporaryArray = (T[])(new Object[sizeOfArray + list.size()]);
+        }
+        for (int i = 0; i < sizeOfArray; i++) {
+            temporaryArray[i] = array[i];
+        }
+        for (int i = sizeOfArray; i < sizeOfArray + list.size(); i++) {
+            temporaryArray[i] = list.get(indexOfArrayList);
+            indexOfArrayList++;
+        }
+        array = temporaryArray;
+        sizeOfArray += list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= sizeOfArray || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MASSAGE);
+        }
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= sizeOfArray || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MASSAGE);
+        }
```

Comment:
Duplication of code, divide this to separate method

Reply:
Done.